# Introduction
This notebook explores information and insights of raw data.
## Importing libraries
NumPy, Pandas and Matplotlib are imported for initial data analysis, extracting statistical insights, and plotting correlations of potential features.

In [3]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt

import os

Now, we read the raw data by using Pandas

In [8]:
icloud_path = os.path.expanduser('~/Library/Mobile Documents/com~apple~CloudDocs/Documents/MyLearning/MachineLearning/Project1/')
os.chdir(icloud_path)
os.getcwd()

'/Users/dushyant/Documents/MyLearning/MachineLearning/Project1'

In [11]:
train_data = pd.read_csv("data/raw/train.csv")
test_data = pd.read_csv("data/raw/test.csv")

In [16]:
train_data.head()

tBodyAcc-mean()-X  tBodyAcc-mean()-Y  tBodyAcc-mean()-Z  tBodyAcc-std()-X  \
0           0.288585          -0.020294          -0.132905         -0.995279   
1           0.278419          -0.016411          -0.123520         -0.998245   
2           0.279653          -0.019467          -0.113462         -0.995380   
3           0.279174          -0.026201          -0.123283         -0.996091   
4           0.276629          -0.016570          -0.115362         -0.998139   

   tBodyAcc-std()-Y  tBodyAcc-std()-Z  tBodyAcc-mad()-X  tBodyAcc-mad()-Y  \
0         -0.983111         -0.913526         -0.995112         -0.983185   
1         -0.975300         -0.960322         -0.998807         -0.974914   
2         -0.967187         -0.978944         -0.996520         -0.963668   
3         -0.983403         -0.990675         -0.997099         -0.982750   
4         -0.980817         -0.990482         -0.998321         -0.979672   

   tBodyAcc-mad()-Z  tBodyAcc-max()-X  ...  fBodyBodyGyroJerkMag-kurtosis()  \
0         -0.923527         -0.934724  ...                        -0.710304   
1         -0.957686         -0.943068  ...                        -0.861499   
2         -0.977469         -0.938692  ...                        -0.760104   
3         -0.989302         -0.938692  ...                        -0.482845   
4         -0.990441         -0.942469  ...                        -0.699205   

   angle(tBodyAccMean,gravity)  angle(tBodyAccJerkMean),gravityMean)  \
0                    -0.112754                              0.030400   
1                     0.053477                             -0.007435   
2                    -0.118559                              0.177899   
3                    -0.036788                             -0.012892   
4                     0.123320                              0.122542   

   angle(tBodyGyroMean,gravityMean)  angle(tBodyGyroJerkMean,gravityMean)  \
0                         -0.464761                             -0.018446   
1                         -0.732626                              0.703511   
2                          0.100699                              0.808529   
3                          0.640011                             -0.485366   
4                          0.693578                             -0.615971   

   angle(X,gravityMean)  angle(Y,gravityMean)  angle(Z,gravityMean)  subject  \
0             -0.841247              0.179941             -0.058627        1   
1             -0.844788              0.180289             -0.054317        1   
2             -0.848933              0.180637             -0.049118        1   
3             -0.848649              0.181935             -0.047663        1   
4             -0.847865              0.185151             -0.043892        1   

   Activity  
0  STANDING  
1  STANDING  
2  STANDING  
3  STANDING  
4  STANDING  

[5 rows x 563 columns]

It shows that are there are 563 features. But how many rows/data points?

In [17]:
train_data.shape[0]

7352

In [18]:
test_data.shape[0]

2947

Now we know that there are 7352 data points from train data and 2947 data points from test data. So what would be the ratio of train-test split?

In [21]:
test_ratio = test_data.shape[0] / (train_data.shape[0] + test_data.shape[0])
print("The test ratio is ", test_ratio)

The test ratio is  0.28614428585299545


There are 563 features, which are very challenging to plot and estimate correlation manually - these include either plotting columns, or using correlation measures for all the features.
Rather, the following approaches could be adopted -
1. Calculating correltion matrix, however, this would also result in a matrix of shape 563 X 563, which is a huge matrix to analyse.
2. Reducing dimensionality for visualisation, either by setting a threshold wherein if a correlation measure is above this threshold, those features are considered. The other option is to sort correlation measure in descending order and consider top-N correlations.
3. Dropping redundant features - by removing highly correlated features.
4. Using dimensionality reduction, such as PCA or t-SNE, to visualise.

In [24]:
# (1): Calculating correlation matrix, and importing the matrix as a separate csv


In [ ]:
# (2)A: Setting threshold
# Get pairs of features with high correlation
high_corr = corr_matrix.abs() > 0.8
np.fill_diagonal(high_corr.values, False)  # exclude self-correlation

# Show which pairs exceed the threshold
high_corr_pairs = high_corr[high_corr].stack()
print(high_corr_pairs)

In [ ]:
# (2)B: Considering top-N correlations

In [ ]:
# (3): Dropping redundant features

In [ ]:
# (4): Using PCA for dimensionality reduction
pca = PCA(n_components=2)
components = pca.fit_transform(df.drop('target', axis=1))

plt.scatter(components[:, 0], components[:, 1], c=df['target'])
plt.xlabel('PC1'); plt.ylabel('PC2'); plt.title('PCA Projection')
plt.show()